File: go-samples/probabilities/Combinatorics.go
package mathfuncs

/*
This package provides operations that can be used in counting and arranging objects in probabilistic mathematics.
*/

/*
Create a 'Combinatorics' struct that will provide bound functions for calculating combinatorial operations. Add a map 'memo' with a cap of 50 to store the results of previously calculated factorials.
*/
type Combinatorics struct {
	memo map[uint]uint
}

/*
Create a function 'NewCombinatorics()' that initializes a 'Combinatorics' object and returns it.	
*/
func NewCombinatorics() Combinatorics {
	return Combinatorics{memo: make(map[uint]uint, 50)}
}

/*
Create a bound function 'Factorial()' that calculates the factorial of a number 'n' and returns the result. It uses memoization to store the results of previously calculated factorials.
Arguments:
 - 'n': the number for which the factorial is to be calculated.
Return:
 - The factorial of 'n'.
Implementation:
- If the Combinatorics object does not exist in the 'memo' map, create it.
- If the factorial of 'n' is not in the 'memo' map, calculate it and store it in the 'memo' map.
- Return the factorial of 'n'.
*/
func (c Combinatorics) Factorial(n uint) uint {
	if c.memo == nil {
		c.memo = make(map[uint]uint, 50)
	}
	if _, ok := c.memo[n]; !ok {
		if n == 0 {
			c.memo[n] = 1
		} else {
			c.memo[n] = n * c.Factorial(n-1)
		}
	}
	return c.memo[n]
}


/*
Create a bound function 'Permutations()'.
Arguments:
- takes two unsigned integer parameters 'n' and 'r'.
- returns the number of permutations of 'n' items taken 'r' at a time.
Definition:
- Permutations are the number of ways to arrange 'r' items from a set of 'n' items.
Implementation:
	return Factorial(n) / Factorial(n - r);
*/
func (c Combinatorics) Permutations(n, r uint) uint {
	return c.Factorial(n) / c.Factorial(n-r)
}

File: python-samples/combinatorics/combinatorics.py
"""
This file contains functions for combinatorics.
"""

"""
Create a function 'factorial' that returns the factorial of a number.
The factorial of a number is the product of all positive integers less than or equal to the number.
Arguments:
n -- an integer
Return:
an integer
Examples:
factorial(-1) => None
factorial(0) => 1
factorial(1) => 1
factorial(2) => 2
factorial(3) => 6
"""
def factorial(n):
    if n == 0 or n == 1:
        return 1
    return n * factorial(n - 1)

"""
Create a function 'permutations' that returns the number of permutations of k elements from a set of n elements.
The number of permutations of k elements from a set of n elements is the number of ways to arrange k elements from a set of n elements.
Arguments:
n -- an integer
k -- an integer
Return:
an integer
Examples:
permutations(0, 0) => 1
permutations(1, 0) => 1
permutations(1, 1) => 1
permutations(2, 1) => 2
permutations(2, 2) => 2
permutations(3, 1) => 3
permutations(3, 2) => 6
"""
def permutations(n, k):
    return factorial(n) // factorial(n - k)







